### PR TITLE
Implement X509Certificate2.RawDataMemory

### LINKED ISSUE #57448: [API Proposal]: X509Certificate2.RawDataMemory
### Background and motivation

The `X509Certificate2` class has a property called `RawData` which returns the raw data of the X.509 certificate object in DER form.

This property however allocates every time it is accessed to return a clone to the caller.

https://github.com/dotnet/runtime/blob/98ca2ffa27ed23d106e7131bac8cb17a46b905d3/src/libraries/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Certificate2.cs#L315-L326

This means code written like:

```C#
Encode(cert.RawData, cert.RawData.Length)
```

or anything else that accesses `RawData` multiple times is going to allocate a several hundred bytes each time the property is accessed. Removing the defensive copy is likely a breaking change.

It would be good if we exposed a way to get the `RawData` without returning a defensive copy every time. In order to do this, the return type must not expose a way of mutating (assuming safe APIs are being used).

### API Proposal

```diff
namespace System.Security.Cryptography.X509Certificates {
    public class X509Certificate2 : X509Certificate {
+       public ReadOnlyMemory<byte> RawDataMemory { get; }
    }
}
```

### API Usage

```C#
ReadOnlyMemory<byte> certData = cert.RawDataMemory;

// or for those that need a span
ReadOnlySpan<byte> certData = cert.RawDataMemory.Span;
```

### Alternatives

I'm not tied to the idea of returning a `ReadOnlyMemory<byte>`. If something else makes more sense, that's fine too.

### Risks

Small risk would be someone using unsafe APIs or `fixed` to coerce the `ReadOnlyMemory<byte>` in to something writable, but I think that is acceptable.

### PR DESCRIPTION
**Pull request by vcsjones**
Closes #57448

### CODE REVIEW

**Comment by vcsjones** [comment]
I didn't go with the custom MemoryManager trick. For one I would probably have a difficult time figuring out how to do it correctly, and for two because it still feels a bit unnecessary to me. `MemoryMarshal` is an "I know what I am doing" type, and even if we did do it, it would not address the pinning possibility.

**Review by bartonjs — APPROVED** [review]

**Comment by GrabYourPitchforks** [comment]
Should we doc that the returned `ROM<byte>` has an independent lifetime from the underlying certificate? `ROM<byte>` is supposed to be an abstraction, but we're returning a property whose correct usage relies on the buffer having an independent lifetime, and people will absolutely depend on this behavior. Might be worthwhile to formalize this expectation in docs.

**Comment by vcsjones** [comment]
@GrabYourPitchforks 

> Should we doc that the returned `ROM<byte>` has an independent lifetime from the underlying certificate? 

What would documentation like that look like? I... think... given the API shape it's a reasonable expectation (otherwise it'd be a `CopyRawDataTo(..)` or something like that.

Would codifying in a test be more reasonable, or is that not strong enough? (I'm starting to think the test should exist regardless... let me add that).

**Comment by GrabYourPitchforks** [comment]
> Would codifying in a test be more reasonable, or is that not strong enough?

That's probably good enough honestly. Get the `ROM<byte>`, dispose the cert, then validate the buffer you previously got is still filled with delicious X.509 goodness.

File: src/libraries/System.Security.Cryptography.X509Certificates/tests/CertTests.cs [hunk 1]
```diff
@@ -474,6 +475,30 @@ public static void CopyResult_RawData()
...
         }
 
+        [Fact]
+        public static void RawDataMemory_NoCopy()
+        {
+            using (X509Certificate2 cert = new X509Certificate2(TestData.MsCertificate))
+            {
+                ReadOnlyMemory<byte> first = cert.RawDataMemory;
+                ReadOnlyMemory<byte> second = cert.RawDataMemory;
+                Assert.True(first.Span == second.Span, "RawDataMemory returned different values.");
```
Comment (stephentoub) [inline src/libraries/System.Security.Cryptography.X509Certificates/tests/CertTests.cs:485]: It's quite rare to see someone use == with spans and have it actually be correct ;-)

**Review by stephentoub — APPROVED** [review]

**Review by GrabYourPitchforks — APPROVED** [review]
Thanks also for the unit test update!

**Pull request merged**